tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +32,430 @@ def test_shouldBeApplicable():
+
+    goal.addDependency(task1)
+    goal.addDependency(task2)
+
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(current)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    plan = goal.isAchievable(fullContext, interp)
+    assert goal.decomposition
+    assert plan is None
+
+
+def test_ApplicableDeps():
+    goal = Pragmatic()
+
+    task = Task()
+    context = Context("C1")
+    wrongContext = Context("C2")
+    current = []
+
+    qc = QualityConstraint(context, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
 
-    def test_aTaskShouldBeAchievable():
-        commonMetrics = CommonMetrics()
-        task = Task()
+    task.addApplicableContext(context)
+    task.setProvidedQuality(context, CommonMetrics.SECONDS, 13)
 
-        currentContext = Context("C1")
-        fullContext = []
-        fullContext.append(currentContext)
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(context)
+    goal.getInterpretation().addQualityConstraint(qc)
 
-        qc = QualityConstraint(currentContext, commonMetrics.SECONDS, 15, 'LESS_OR_EQUAL_TO')
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
 
-        task.addApplicableContext(currentContext)
-        task.setProvidedQuality(currentContext, commonMetrics.SECONDS, 12)
+    current.append(wrongContext)
+    assert goal.isAchievable(current, interp) is None
 
-        interp = Interpretation()
-        interp.addQualityConstraint(qc)
+    current.append(context)
+    assert len(goal.isAchievable(current, interp).getTasks()) == 1
 
-        assert True == task in task.isAchievable(fullContext, interp).getTasks()
 
-    def test_shouldAddSeveralContextsAtOnce():
-        context1 = Context("C1")
-        context2 = Context("C2")
+def testGetApplicableQC():
+    goal = Pragmatic(Decomposition.AND)
 
-        task = Task()
-        originalSize = task.getApplicableContext().len()
-        set = []
+    task = Task()
+    context = Context("C1")
+    anotherContext = Context("C2")
+
+    fullContext = []
+
+    qc = QualityConstraint(context, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    stricter = QualityConstraint(
+        anotherContext, CommonMetrics.SECONDS, 10, Comparison.LESS_OR_EQUAL_TO)
+
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(context)
+    goal.getInterpretation().addQualityConstraint(qc)
+    goal.getInterpretation().addQualityConstraint(stricter)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    fullContext.append(context)
+    # assertEquals(null, goal.isAchievable(fullContext, interp))
+
+    assert qc in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+
+    fullContext.append(anotherContext)
+    assert qc in goal.getInterpretation().getQualityConstraints(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)